File: Greedy/ChooseAndSwap/ChooseAndSwap.js
function chooseAndSwap(text){
    let arr = new Array(26);
    arr.fill(-1);
    let i,len=text.length,ind,j;
    for(i=0;i<len;i++){
        let c = text[i];
        ind = c.charCodeAt()-'a'.charCodeAt();
        if(arr[ind]==-1){
            arr[ind] = i;
        }
    }
    let char,char2;
    let bool = false;
    
    for(i=0;i<len;i++){
        j=text[i].charCodeAt();
        while(j>=97){
            ind = j;
            if(arr[ind-97]>=-1 && arr[ind-97]>i){
                char = text[i];
                char2 = String.fromCharCode(ind);
                bool = true;
            }
            j--;
        }
        if(bool){
            break;
        }
    }
    console.log('char is ', char,', char2 is ', char2);
    for(i=0;i<len;i++){
        if(text[i]==char){
            text[i]=char2;
        }else if(text[i]==char2){
            text[i]=char;
        }
    }
    console.log(arr);
    console.log(text);
}

chooseAndSwap('ccab'.split(''));

File: Greedy/ChooseAndSwap/ChooseAndSwap.go
package main

import (
	"fmt"
)

func method1(str []byte) {
	arr := [26]int{}
	for i := 0; i < 26; i++ {
		arr[i] = -1
	}
	var char1 byte
	var char2 byte
	char1 = 'a'
	char2 = 'a'
	n := len(str)

	for i := 0; i < n; i++ {
		if arr[str[i]-97] == -1 {
			arr[str[i]-97] = i
		}
	}
	fmt.Println("arr is ", arr)
outer:
	for i := 0; i < n; i++ {
		temp := byte(97)
		for ; temp < str[i]; temp++ {
			if arr[temp-97] >= 0 && arr[(temp-97)] > i {
				char1 = temp
				char2 = str[i]
				break outer
			}
		}
	}
	fmt.Println("char1 is ", char1, ", char2 is ", char2)
	for i := 0; i < n; i++ {
		if str[i] == char1 {
			str[i] = char2
		} else if str[i] == char2 {
			str[i] = char1
		}
	}
	fmt.Println("str is ", str)
}

func main() {
	str := "ccab"
	arr := make([]byte, len(str))
	for i := 0; i < len(str); i++ {
		arr[i] = str[i]
	}
	fmt.Println(arr)
	method1(arr)
	fmt.Println(arr)
}
